flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -717,14 +717,11 @@ public String toString() {
                 waitAndGetNextData(blocking);
         if (!next.isPresent()) {
             throughputCalculator.pauseMeasurement();
-            getAvailableFuture()
-                    .whenComplete(
-                            (future, ex) -> {
-                                throughputCalculator.resumeMeasurement();
-                            });
             return Optional.empty();
         }
 
+        throughputCalculator.resumeMeasurement();
+
```

Comment:
Moving this code here is changing the following things:

1. If the task is back pressured, we won't pull any buffers until back pressure goes away, so the measurement will resume later (potentially much later)
2. This problem is exasperated if we have many `SingleInputGate`s.

Reply:
We pause the calculation when the task is idle rather than back pressured. The only problem that I see here is if we have a lot of `SingleInputGates` and the idly is pretty fast switched to backpressure. So in this case some of `SingleInputGates` can be resumed a little later than in the old implementation. But in general, I don't think that it is so big a problem for calculation.
There is the idea to resume calculation via mailboxProcessor which sounds good for but unfortunately, it is too difficult to implement (a lot of refactoring requires) and it doesn't provide a lot of advantages compared to the current solution. So I think that it makes sense to leave the solution as is.